### PR TITLE
FEAT: Display Token Usage in the UI

### PR DESCRIPTION
**Pull request by tahierhussain**
## What

1. API to fetch the token usage data.
2. UI changes to display the token usage data (format - `Total Token Count`).

-

## Why

The token usage for prompt runs was not being displayed to the user earlier.

-

## How

NA

-

## Can this PR break any existing features. If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)

Yes, there are quite a few changes in multiple components, so there is a possibility that this PR may affect other features.

-

## Database Migrations
NA
- 

## Env Config
NA
- 

## Relevant Docs
NA
-

## Related Issues or PRs
https://github.com/Zipstack/unstract-cloud/pull/104
-

## Dependencies Versions
NA
-

## Notes on Testing
NA
-

## Screenshots
image

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

**Comment by hari-kuriakose** [comment]
@tahierhussain I assume the token usage is in the format `<embedding>|<prompt>|<completion>|<total>`. Can we please add this to the PR description?

1. Is the total tokens usage calculation correct? From the screenshot, usage of both prompts seems to be less by 1.
    e.g. `1 + 808 + 3` = `811` instead of `812`

2. Are we showing a tooltip with explanation of display format on hover over token usage?

**Comment by tahierhussain** [comment]
> @tahierhussain I assume the token usage is in the format `<embedding>|<prompt>|<completion>|<total>`. Can we please add this to the PR description?
> 
> 1. Is the total tokens usage calculation correct? From the screenshot, usage of both prompts seems to be less by 1.
>    e.g. `1 + 808 + 3` = `811` instead of `812`
> 2. Are we showing a tooltip with explanation of display format on hover over token usage?

@hari-kuriakose I have updated the PR description.

1. Thanks for the catch. You're right. The calculation is not correct. @gaya3-zipstack @Deepak-Kesavan Looks like the total value is not including the embedding tokens. I have verified it by checking the db. Can you please check this?

2. Yes, a tooltip is added which explains the format.

**Comment by gaya3-zipstack** [comment]
> > @tahierhussain I assume the token usage is in the format `<embedding>|<prompt>|<completion>|<total>`. Can we please add this to the PR description?
> > 
> > 1. Is the total tokens usage calculation correct? From the screenshot, usage of both prompts seems to be less by 1.
> >    e.g. `1 + 808 + 3` = `811` instead of `812`
> > 2. Are we showing a tooltip with explanation of display format on hover over token usage?
> 
> @hari-kuriakose I have updated the PR description.
> 
> 1. Thanks for the catch. You're right. The calculation is not correct. @gaya3-zipstack @Deepak-Kesavan Looks like the total value is not including the embedding tokens. I have verified it by checking the db. Can you please check this?
> 2. Yes, a tooltip is added which explains the format.

@tahierhussain @hari-kuriakose @Deepak-Kesavan 
On point (1), looking at the token usage logging, total_tokens is a field which seems to be applicable only to LLM tokens and not embedding tokens as shown in the image below. 
If we need to add the embedding token also to the total then we need to change the [SDK code](https://github.com/Zipstack/unstract-sdk/blob/0e76774ac9f15fbbf410db6fd25a1fc7f257b196/src/unstract/sdk/audit.py#L82) to do so .
My understanding is that users are only concerned about LLM token count. So do we really need to change "total count" to "total LLM count" instead?

image

**Review by gaya3-zipstack — COMMENTED** [review]
Token usage API handler seems fine. May have to update changes related to total token count depending on the feedback.

File: frontend/src/components/custom-tools/prompt-card/PromptCard.jsx [hunk 1]
```diff
@@ -411,11 +418,26 @@ function PromptCard({
   };
 
   const handleRunApiRequest = async (docId) => {
+    const runId = generateUUID();
```
Comment (gaya3-zipstack) [inline frontend/src/components/custom-tools/prompt-card/PromptCard.jsx:421]: @tahierhussain There were 2 APIs where this runId may be required. 
1. fetch_response
2. single_pass_extraction
Is this handled in both places?

Comment (gaya3-zipstack) [inline frontend/src/components/custom-tools/prompt-card/PromptCard.jsx:421]: @tahierhussain There were 2 APIs where this runId may be required. 
1. fetch_response
2. single_pass_extraction
Is this handled in both places?

Comment (tahierhussain) [reply frontend/src/components/custom-tools/prompt-card/PromptCard.jsx:421]: Yes, @gaya3-zipstack. It is handled in the cloud PR.

**Comment by Deepak-Kesavan** [comment]
@tahierhussain Can you update the screenshot with latest changes?

**Review by gaya3-zipstack — APPROVED** [review]
Changes look fine.

**Comment by tahierhussain** [comment]
> @tahierhussain Can you update the screenshot with latest changes?

The current screenshot is the updated one.

**Review by Deepak-Kesavan — APPROVED** [review]

**Comment by hari-kuriakose** [comment]
> > > @tahierhussain I assume the token usage is in the format `<embedding>|<prompt>|<completion>|<total>`. Can we please add this to the PR description?
> > > 
> > > 1. Is the total tokens usage calculation correct? From the screenshot, usage of both prompts seems to be less by 1.
> > >    e.g. `1 + 808 + 3` = `811` instead of `812`
> > > 2. Are we showing a tooltip with explanation of display format on hover over token usage?
> > 
> > 
> > @hari-kuriakose I have updated the PR description.
> > 
> > 1. Thanks for the catch. You're right. The calculation is not correct. @gaya3-zipstack @Deepak-Kesavan Looks like the total value is not including the embedding tokens. I have verified it by checking the db. Can you please check this?
> > 2. Yes, a tooltip is added which explains the format.
> 
> @tahierhussain @hari-kuriakose @Deepak-Kesavan On point (1), looking at the token usage logging, total_tokens is a field which seems to be applicable only to LLM tokens and not embedding tokens as shown in the image below. If we need to add the embedding token also to the total then we need to change the [SDK code](https://github.com/Zipstack/unstract-sdk/blob/0e76774ac9f15fbbf410db6fd25a1fc7f257b196/src/unstract/sdk/audit.py#L82) to do so . My understanding is that users are only concerned about LLM token count. So do we really need to change "total count" to "total LLM count" instead?
> 
> image

@gaya3-zipstack @tahierhussain Showing the token usage for LLM alone is good enough.
We can show the total alone for now, but maybe a break down can be provided in the tooltip later.

**Pull request merged**